index_part.go

```diff
@@ -72,63 +86,27 @@ func (j fieldIndexer) Name() string {
 	return j.path
 }
 
-func (j fieldIndexer) Keys(document Document) ([]Key, error) {
-	// first get the raw values from the query path
-	rawKeys, err := document.ValuesAtPath(j.path)
-	if err != nil {
-		return nil, err
-	}
-
-	// run the tokenizer
-	tokenized := make([]interface{}, 0)
-	if j.tokenizer == nil {
-		tokenized = rawKeys
-	} else {
-		for _, rawKey := range rawKeys {
-			tokens := j.Tokenize(rawKey)
-			tokenized = append(tokenized, tokens...)
-		}
-	}
-
-	// run the transformer
-	transformed := make([]interface{}, len(tokenized))
-	if j.transformer == nil {
-		transformed = tokenized
-	} else {
-		for i, rawKey := range tokenized {
-			transformed[i] = j.transformer(rawKey)
-		}
-	}
-
-	// to Keys
-	keys := make([]Key, len(transformed))
-	for i, t := range transformed {
-		keys[i], err = toBytes(t)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	return keys, nil
+func (j fieldIndexer) Path() string {
+	return j.path
 }
 
-func (j fieldIndexer) Tokenize(value interface{}) []interface{} {
+func (j fieldIndexer) Tokenize(scalar Scalar) []Scalar {
 	if j.tokenizer == nil {
-		return []interface{}{value}
+		return []Scalar{scalar}
 	}
 
-	if s, ok := value.(string); ok {
+	if s, ok := scalar.value.(string); ok {
 		tokens := j.tokenizer(s)
-		result := make([]interface{}, len(tokens))
+		result := make([]Scalar, len(tokens))
 		for i, t := range tokens {
-			result[i] = t
+			result[i] = ScalarMustParse(t)
```

Comment:
wouldn't it make more sense to make `Scalar` an interface and have implementing structs for each supported type?

Reply:
changed it, leads to more code but less switch statements